lib/index.js

```diff
@@ -0,0 +1,300 @@
+ *         ...nextPayload.obj,
+ *         traits: newOrUpdatedTraits || nextPayload.obj.traits,
+ *       },
+ *     };
+ *     // do something smart with `debouncedPayload`
+ *   }
+ * ```
+ */
+export const debouncePayloads = (
+  previousPayload,
+  nextPayload,
+  getPayloadProperty,
+) => {
+  let newOrUpdatedTraits = null;
+  if (!_isNil(nextPayload) && !_isNil(previousPayload)) {
+    const nextAnonymousId = getPayloadProperty(nextPayload, 'anonymousId');
+    const previousAnonymousId = getPayloadProperty(
+      previousPayload,
+      'anonymousId',
+    );
+    const nextUserId = getPayloadProperty(nextPayload, 'userId');
+    const previousUserId = getPayloadProperty(previousPayload, 'userId');
+    if (
+      nextAnonymousId !== previousAnonymousId ||
+      nextUserId !== previousUserId
+    ) {
+      return { nextPayload, newOrUpdatedTraits };
+    }
+
+    const nextTraits = getPayloadProperty(nextPayload, 'traits');
+    const previousTraits = getPayloadProperty(previousPayload, 'traits');
+    newOrUpdatedTraits = diffObject(nextTraits, previousTraits || {}, {
+      diffArray: false,
+    });
+    if (_isNil(newOrUpdatedTraits)) {
+      return { nextPayload: null, newOrUpdatedTraits };
+    }
+
+    // Braze charges for every trait sent to it even if that trait isn't
+    // changed. So we only want to include traits that have changed in the
+    // final payload.
+    return { nextPayload, newOrUpdatedTraits };
+  }
+  return { nextPayload, newOrUpdatedTraits };
+};
+
+/**
+ * `debouncePayloadSync` provides a full solution that integrates with the
+ * Analytics.js Source Middleware to debounce data being before it is sent
+ * to the Braze destination. It stores previous payloads in `localStorage`,
+ * or a similar solution which you can override, and compares the `payload`
+ * provided by this middleware against the previous versions to only send
+ * new or updated `traits`.
+ * @param {Object} payload The data provided by the Analytics.js middleware
+ * @param {Function} options.fetchPayload An optional function to override `localStorage.getItem`
+ * @param {Function} options.storePayload An optional function to override `localStorage.setItem`
+ * @return {Object} The debounced payload to send to Braze. If `null`,
+ * there is no new or updated data to send.
+ * @example
+ * ```
+ *   const _identifyDebounceSourceMiddleware = ({ payload, next, integrations }) => {
+ *     // TODO filter Braze integration, called `AppBoy`
+ *     if (payload.type() !== 'identify') {
+ *       next(payload);
+ *       return;
+ *     }
+ *     const identifyPayload = debouncePayloadSync(payload);
+ *     if (identifyPayload) {
+ *       next(identifyPayload);
+ *     }
+ *   }
+ *   analytics.addSourceMiddleware(_identifyDebounceSourceMiddleware);
+ * ```
+ *
+ */
+export const debouncePayloadSync = (
+  payload,
+  {
+    fetchPayload = _get(global, 'localStorage.getItem'),
+    persistPayload = _get(global, 'localStorage.setItem'),
```

Comment:
Same nit here: this works because we're defensively assuming `localStorage` might not exist in the execution environment on the `global` object _but_ we could additionally provide a separate entry point for organizational purposes so that this becomes `import { debouncePayloadSync } from 'braze-segment-debounce/browser';`

Reply:
Or even make these options required here (since it being synchronous isn't necessarily a browser-only thing and that other environments like react native might also want a synchronous function to use) and then the `browser.js` entrypoint would just be like (pseudocode)

```js
import { debouncePayloadSync } from '.';

export const debouncePayloadSyncViaLocalStorage = (payload, options = {
  fetchPayload = (...args) => localStorage.getItem(...args),
  persistPayload = (...args) => localStorage.setItem(...args),
}) {
  return debouncePayloadSync(payload, options);
}
```